### PR TITLE
[FlashInfer] Add SM12x b12x NVFP4 MoE support (Laguna S 2.1 + DFlash working in NVFP4 on SM121)

### PR DESCRIPTION
**Pull request by robbiemu**
## What this is

This PR adds the missing SGLang adaptation between compressed-tensors NVFP4 MoE checkpoints and FlashInfer's SM12x `b12x_fused_moe` kernel.

With [#32062](https://github.com/sgl-project/sglang/pull/32062), the FlashInfer `0.6.15.post1` upgrade from [#31927](https://github.com/sgl-project/sglang/pull/31927), and this PR, `poolside/Laguna-S-2.1-NVFP4` runs with `poolside/Laguna-S-2.1-DFlash-NVFP4` on a DGX Spark (GB10 / SM121 / TP=1).

The historical target checkpoint passed all 12 semantic checks in the original Laguna+DFlash qualification regime: greedy output-set parity, native tool calls, speculative-acceptance invariants, continuation, cached-prefix reuse, a 1,498-token prompt, and an above-window long prompt. DFlash proposal and acceptance counters confirmed that speculative decoding was active rather than silently falling back to target-only decoding.

Poolside subsequently replaced the target weights. The replacement checkpoint has now also been tested from the exact source at this PR head. Target-only requests completed through 127,264 prompt tokens, and the documented DFlash configuration completed semantic requests through 127,264 prompt tokens and exact needle-retrieval capacity probes through 180,174 prompt tokens.

The source change is deliberately small:

- select B12x for compressed-tensors NVFP4 MoE on SM120/SM121 when the TRTLLM backend is not selected;
- reorder FC1 from checkpoint `[w1, w3]` order to the kernel's `[w3, w1]` order;
- convert the already-swizzled block scales to the six-dimensional MMA layout expected by B12x;
- adapt the checkpoint's W13, W2, and FC2 scale contract to the legacy B12x API; and
- dispatch through `b12x_fused_moe` while preserving the existing FlashInfer CUTLASS path elsewhere.

The implementation is generic to compressed-tensors NVFP4 MoE on SM12x. Laguna is the model and architecture combination qualified here; this PR does not claim validation for every compressed-tensors NVFP4 model.

This is a stacked PR:

- [#32062](https://github.com/sgl-project/sglang/pull/32062) supplies the required Laguna attention-planning fix. Until it merges, GitHub will show that PR's files and commits here as well.
- [#31927](https://github.com/sgl-project/sglang/pull/31927), now merged into `main`, supplies FlashInfer `0.6.15.post1`. This PR intentionally does not duplicate that dependency upgrade.
- The new MoE work is the final commit and changes only `flashinfer_cutlass.py` and `compressed_tensors_w4a4_nvfp4_moe.py`.

Qualified revisions and runtime:

- historical target: `poolside/Laguna-S-2.1-NVFP4` at `b482b5d57fda6e4e562a652869bde24ba2a57c92`
- replacement target: `poolside/Laguna-S-2.1-NVFP4` at `07614121b31898586430f189d27a25a0be310843`
- drafter: `poolside/Laguna-S-2.1-DFlash-NVFP4` at `723794750422b3efbf3a7b3af76dffb4ba035943`
- SGLang PR head: `9b9002f4a50e0ba0399b65c5e8fa6acf3b3c1301`
- DGX Spark GB10 / SM121 / TP=1
- page size 1, chunked prefill 2,048, DFlash block size 16
- target-only long-context runs: FP8 E4M3 KV, FlashInfer attention, CUDA graphs disabled
- DFlash long-context runs: BF16 KV, FlashInfer target and draft attention, CUDA graphs and overlap scheduling enabled
- FlashInfer `0.6.15.post1`
- models loaded from the standard Hugging Face cache and mounted read-only

Checks run on the exact minimized two-file diff:

```text
git diff --check: passed
py_compile: passed
isort --check-only: passed
black --check: passed
```

The exact minimized PR source was rebuilt into the image used for the replacement-checkpoint short- and long-context runs. No throughput or latency claim is made.

### Replacement-checkpoint qualification

The matched parent-control image retained the attention fix and FlashInfer upgrade but restored the two MoE files to their parent versions. It produced corrupt text, logits, and tool output on the replacement checkpoint. This PR therefore remains necessary with the replacement weights.

| Mode | Authoritative prompt token count(s) | Result |
| --- | ---: | --- |
| Target only | 60,010 / 59,969 | Exact needle retrieval / arithmetic numerically correct but violating the requested output format |
| Target only | 127,264 | Exact needle retrieval |
| Target only | 127,223 | Exact arithmetic output `1517` |
| DFlash | 29,804 / 29,763 | Exact needle retrieval / exact arithmetic |
| DFlash | 60,010 / 59,969 | Exact needle retrieval / arithmetic numerically correct but violating the requested output format |
| DFlash | 127,264 | Exact needle retrieval |
| DFlash | 127,223 | Exact arithmetic output `1517` |
| DFlash capacity probe | 163,784 | Exact needle retrieval |
| DFlash capacity probe | 180,174 | Exact needle retrieval |

DFlash proposal and acceptance counters independently confirmed speculative execution at the 32K, 64K, and corrected 128K rungs. The 180,174-token capacity probe reached 8.67 GiB minimum host `MemAvailable`, close to the approved 8 GiB safety floor, so no higher DFlash request was attempted. No 262K+DFlash run was made.

The long inputs used fixed repeat counts with no client-side tokenization. Completed-response `usage.prompt_tokens` values are authoritative. No completed long-context request produced an incorrect arithmetic result, failed needle retrieval, or a new verification loop.

The earlier paired 15-run short-context suite did, however, produce the same three arithmetic verification loops on the historical and replacement target revisions. The replacement checkpoint did not eliminate that behavior under the tested NVFP4/B12x path. The long-context results demonstrate broad functional execution; they do not establish that every agentic or formatting defect is resolved.

## Minimum working configuration from ablation

The table below shows the complete tested configuration in each row. **Only the fourth row works.** The result is not “the dependency upgrade or this PR”; the working configuration requires the attention fix, the newer FlashInfer kernel, and this PR together.

| Complete tested configuration | Laguna target + DFlash | Why |
| --- | --- | --- |
| SGLang main | Does not run correctly | Both attention planning and the SM121 NVFP4 MoE integration are incomplete. |
| Main + attention PR #32062 | Does not run correctly | The model's mixed query-head/window metadata is planned correctly, but SGLang still does not call the SM121 NVFP4 MoE kernel with the checkpoint's required layouts and scales. |
| FlashInfer-upgrade PR #31927 + attention PR #32062 | Does not run correctly | The B12x kernel is available, but a dependency upgrade cannot reorder the checkpoint weights, transform SGLang's scale layout, or translate its scale contract. |
| FlashInfer-upgrade PR #31927 + attention PR #32062 + this PR | **Works** | All three layers are present: correct attention planning, a usable B12x kernel, and the SGLang-side MoE adaptation. The historical target passed the original 12-check suite; the replacement target passed the qualification summarized above. |

Starting from the working fourth row, component ablations established the minimum inside this PR:

- **SM121 B12x dispatch:** without it, compressed-tensors NVFP4 remains on a kernel path that does not support this model correctly on GB10.
- **Six-dimensional MMA scale layout:** B12x does not consume SGLang's existing flattened/swizzled scale view directly.
- **FC1 `[w1, w3]` to `[w3, w1]` reorder:** the checkpoint and B12x gated-activation conventions use opposite concatenation order.
- **W13 scale folding and neutral W13 alpha as a pair:** the legacy API gives one alpha two responsibilities. Changing only the block scale or only the alpha double-applies or omits part of the checkpoint scale.
- **W2 scale folding and neutral W2 alpha as a pair:** FC2 has the same weight-global-scale mismatch and likewise requires both sides of the translation.
- **Neutral FC2 input scale:** B12x dynamically quantizes the FC2 input; retaining the checkpoint's static FC2 input scale applies the wrong contract.

Removing any one of those behaviors broke the working configuration. Removing only one side of either normalization pair also failed. The unrelated TRTLLM refactor in the original experimental patch was never on Laguna's executed backend and is not part of this minimum.

## History

The investigation began with stock SGLang, where Laguna failed in both attention planning and NVFP4 MoE execution. The attention failure was isolated first: Laguna's full-attention and sliding-window groups have different query-head counts, while the FlashInfer planner used model-global metadata and shared mutable ragged-planner state. That became the separate, generic attention fix in [#32062](https://github.com/sgl-project/sglang/pull/32062).

Attention correctness did not make the model run. A working vLLM deployment on the same architecture showed that SM121 support used FlashInfer's B12x MoE path. Differential analysis against [vLLM #40082](https://github.com/vllm-project/vllm/pull/40082) identified the missing SGLang-side weight ordering, scale layout, and scale-contract adaptation.

The first working local patch was intentionally broad because it was being used to find the path, not yet to propose upstream code. It combined attention experiments, B12x support, scale handling, and a TRTLLM refactor. That proved a working configuration existed, but not which changes were responsible.

We then rebuilt from clean lineages and tested the dependency upgrade, attention PR, and MoE patch separately. After the four configuration tests above, a component-by-component ablation removed or neutralized each behavior inside the working MoE patch. That reduced the upstream change to the two files in this PR and excluded the TRTLLM refactor and every unrelated attention experiment.

Poolside then published replacement NVFP4 target weights. We downloaded and verified the complete replacement snapshot, repeated the PR/no-PR control, compared all checkpoint scale inputs, reran paired short-context tests, and extended exact-source qualification to long target-only and DFlash contexts. The replacement weights changed model behavior in some prompts but did not remove this PR's necessity or the legacy B12x scale-folding limitation.

<details>
<summary>Other findings from the investigation</summary>

- On SM121/TP=1, relaxing the TRTLLM MHA prefill gate reached the kernel, which then reported `Unsupported architecture`. The observed failure was therefore not the documented TP>=4 Python rejection and could not be solved by widening that guard.
- Upgrading FlashInfer alone did not fix either the attention metadata supplied by SGLang or the checkpoint-to-B12x MoE adaptation.
- DFlash `speculative_num_steps=16` produces 15 draft proposals plus the verifier/bonus position; changing it to 15 was an off-by-one configuration change and was reverted.
- The tested runtime exposes one global KV-cache dtype, so target and drafter use the same KV dtype in a DFlash run.
- DFlash currently rejects grammar/structured-output constraints and logprob requests independently of this PR.
- The replacement target completed target-only semantic requests through 127K. DFlash completed semantic requests through 127K and exact needle-retrieval capacity probes through 180K.
- Target-only and DFlash runs produced the same set of greedy variants in the original qualification. A single-output hash comparison initially looked like corruption, but repeated output-set comparison did not support that conclusion.

</details>

### Current path: implemented here and currently active in vLLM

Our understanding is that vLLM's current merged B12x integration in [vLLM #40082](https://github.com/vllm-project/vllm/pull/40082) uses the same legacy scale contract implemented by this PR: tiny per-expert FP32 weight multipliers are folded into E4M3 block scales, and the corresponding runtime alphas are neutralized. This PR deliberately reproduces that active vLLM approach in SGLang. It makes Laguna operational, but it is not numerically ideal.

A full scan of all 36,096 expert projection groups across both target revisions found effectively identical folding behavior:

- 89.139% of nonzero effective block scales become E4M3 subnormals;
- 0.02504% round to zero;
- mean relative rounding error is 4.571%;
- individual entries reach 100% relative error; and
- some projections round as many as 46.62% of their effective scales to zero.

This mechanism exists in target-only inference as well as DFlash. The replacement and historical checkpoints also produced the same three arithmetic verification-loop failures in the paired short-context suite.

The replacement checkpoint nevertheless completed exact long-context needle retrieval and numerically correct arithmetic across the tested target-only and DFlash rungs. The remaining observed failures were output-contract defects, not incorrect arithmetic or retrieval. These experiments do not causally attribute the looping or formatting defects to scale folding because every usable SGLang path tested here uses the same folded B12x contract.

### Proposed replacement: not implemented by this PR or current vLLM

The following paired approach is different from the implementation in this PR and from vLLM's currently active path. It would replace the scale folding above with a decoupled scale contract:

1. [FlashInfer #3932](https://github.com/flashinfer-ai/flashinfer/pull/3932) corrects subnormal E4M3 decoding and adds a separate `input_global_scale`. That allows the weight multiplier to remain FP32 instead of being baked into block scales.
2. Draft [vLLM #48536](https://github.com/vllm-project/vllm/pull/48536) adopts that API and stops baking `weight_scale_2` into the block scales.
3. SGLang could adopt the same contract while retaining the B12x dispatch, weight reorder, and six-dimensional layout established by this PR.

That is not yet a complete replacement for the current solution:

- the FlashInfer change is open and the vLLM integration is still a draft;
- they are not available as a released, paired FlashInfer+integrator stack;
- the reported Qwen measurements substantially reduced excess reasoning-token usage but did not eliminate the gap; and
- those measurements are not a Laguna result, so a controlled target-only A/B is still required before evaluating DFlash on top.

Laguna's checkpoints contain blockwise E4M3 scales and separate per-expert weight/input global scales. No separately published activation/calibration profile is available to independently audit the calibration; absence of that profile does not mean the checkpoints lack scale data. The model card's sampling guidance is a separate confounder from the kernel-scale problem.

This PR records the minimum known-working SGLang path while making the numerical limitation explicit. The most direct next experiment is to build FlashInfer #3932, port the decoupled-scale integration to SGLang, and compare legacy B12x with decoupled B12x using matched layer/kernel outputs, first-token/logit parity, and the same target-only semantic suite before adding DFlash.

---
### CI States

Latest PR Test (Base): :x: [Run #30305550346](https://github.com/sgl-project/sglang/actions/runs/30305550346)
Latest PR Test (Extra): :x: [Run #30305550506](https://github.com/sgl-project/sglang/actions/runs/30305550506)

### CODE REVIEW

**Comment by robbiemu** [comment]
### Replacement checkpoint qualification update

This supersedes the earlier qualification warning for Poolside's replacement target checkpoint.

The replacement target has now been tested using the exact draft-PR source and dependency stack:

- target: [`poolside/Laguna-S-2.1-NVFP4@07614121b31898586430f189d27a25a0be310843`](https://huggingface.co/poolside/Laguna-S-2.1-NVFP4/commit/07614121b31898586430f189d27a25a0be310843)
- drafter: [`poolside/Laguna-S-2.1-DFlash-NVFP4@723794750422b3efbf3a7b3af76dffb4ba035943`](https://huggingface.co/poolside/Laguna-S-2.1-DFlash-NVFP4/commit/723794750422b3efbf3a7b3af76dffb4ba035943)
- SGLang PR head: `9b9002f4a50e0ba0399b65c5e8fa6acf3b3c1301`
- DGX Spark GB10 / SM121 / TP=1
- FlashInfer `0.6.15.post1`

The replacement checkpoint still requires this PR's SM12x B12x adaptation. A matched image containing the attention fix and FlashInfer upgrade but not this PR produced corrupt text, logits, and tool output.

#### Long-context results

Target-only inference completed:

- exact needle retrieval at 60,010 prompt tokens;
- exact needle retrieval at 127,264 prompt tokens;
- numerically correct arithmetic at approximately 60K, with an output-format violation; and
- exact arithmetic output `1517` at 127,223 prompt tokens.

The documented DFlash configuration completed:

- exact needle retrieval at 29,804, 60,010, and 127,264 prompt tokens;
- correct arithmetic at all three tested context rungs, with the same formatting-only defect at approximately 60K; and
- exact capacity-probe needle retrieval at 163,784 and 180,174 prompt tokens.

DFlash proposal and acceptance counters independently confirmed speculative execution at the 32K, 64K, and corrected 128K rungs. No 262K+DFlash run was attempted. The 180,174-token run reached 8.67 GiB minimum host `MemAvailable`, close to the approved 8 GiB safety floor, so testing stopped there.

Long inputs were constructed with fixed repeat counts and no client-side tokenization. SGLang's completed-response `usage.prompt_tokens` values are the authoritative lengths.

No completed long-context request produced an incorrect arithmetic result, failed needle retrieval, or a new verification loop. However, the earlier paired 15-run short-context suite still produced the same three arithmetic verification loops on both the historical and replacement target revisions. The new checkpoint therefore did not eliminate that previously observed behavior under the tested NVFP4/B12x path.

#### Scale-contract assessment

A full comparison of both checkpoint revisions found effectively identical legacy B12x folding behavior:

- 89.139% of nonzero effective block scales become E4M3 subnormals;
- 0.02504% round to zero;
- mean relative rounding error is 4.571%;
- individual entries reach 100% relative error; and
- some projections round as many as 46.62% of their effective scales to zero.

The numerical warning therefore still applies to the replacement checkpoint. The new functional results substantially strengthen the execution evidence, but they do not establish that scale folding causes the remaining looping or formatting defects because every usable SGLang path tested here uses the same folded B12x contract. A matched decoupled-scale A/B remains the appropriate causal test.
